<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title></title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>



<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<pre><code class="r">library(ProM)
library(Seurat)
</code></pre>

<p>This R package contains ProM algorithm and wrap-up of a few other existing deconvolution algorithms for comparison purpose. The input of the ProM algorithm is a scRNAseq dataset (Seurat object) and the bulk RNAseq data in TPM measurement(gene by sample matrix). The R package contains toy examples of input datasets which can be checked as following.</p>

<pre><code class="r">data(sc)
class(sc)
#&gt; [1] &quot;Seurat&quot;
#&gt; attr(,&quot;package&quot;)
#&gt; [1] &quot;Seurat&quot;
dim(sc)
#&gt; [1] 24383  2021
data(bulk)
class(bulk)
#&gt; [1] &quot;matrix&quot;
dim(bulk)
#&gt; [1] 24062    14
</code></pre>

<p>For each cell in Seurat object, three input features are needed: the minor and major cell cluster each cell was assigned to, and the sample where each cell comes from. The minor cell clusters are the ones that ProM will deconvolute the bulk RNAseq into. The major cell cluster annotations are needed to facilitate ProM to derive cell markers for similar minor clusters. It is fine if some major cell clusters only consist of one minor cell cluster.</p>

<pre><code class="r">table(sc$majorlabel)
#&gt; 
#&gt; Endothelial  Epithelial       T-CD4 
#&gt;         821         400         800
table(sc$minorlabel)
#&gt; 
#&gt;                       Basal      Endothelial_ACKR1_HEVs 
#&gt;                         200                         200 
#&gt;   Endothelial_CA4_capillary        Endothelial_ESM1_tip 
#&gt;                         174                         200 
#&gt;  Endothelial_FBLN5_arterial Endothelial_PROX1_lymphatic 
#&gt;                         200                          47 
#&gt;                     Luminal                 T-CD4_IL17A 
#&gt;                         200                         200 
#&gt;                 T-CD4_PDCD1            T-CD4-naive_CCR7 
#&gt;                         200                         200 
#&gt;               T-CD4-TRM_JUN 
#&gt;                         200
table(sc$sampleID)
#&gt; 
#&gt; SI_357 SI_358 SI_359 SI_360 SI_394 SI_422 SI_423 SI_424 SI_434 SI_435 
#&gt;     85     50     33     58      3      4      2    180    200     14 
#&gt; SI_453 SI_454 SI_456 SI_457 SI_584 SI_585 SI_595 SI_596 SI_603 SI_604 
#&gt;     18    210     18    204    235     28     55      9      9      9 
#&gt; SI_613 SI_651 SI_656 SI_659 SI_696 SI_697 
#&gt;      8    358     87     44     72     28
</code></pre>

<p>The first step is to use prepareInput() function to preprocess the above input data to generate needed data for the ProM deconvolution function. It includes getting the common genes between bulk and single cell datasets, renormalization, getting markers for each cell cluster, calculating reference profiles with gene mean and variance. This may take a few minutes.</p>

<pre><code class="r">d&lt;-prepareInput(bulk,sc,majorlabel = &quot;majorlabel&quot;, minorlabel = &quot;minorlabel&quot;, samplelabel = &quot;sampleID&quot;)
#&gt; [1] &quot;There are 24062 common genes:renormaliation&quot;
#&gt; [1] &quot;calculate major markers&quot;
#&gt; Calculating cluster Epithelial
#&gt; Calculating cluster Endothelial
#&gt; Calculating cluster T-CD4
#&gt; [1] &quot;calculate minor markers&quot;
#&gt; [1] &quot;Epithelial&quot;
#&gt; Calculating cluster Luminal
#&gt; Calculating cluster Basal
#&gt; [1] &quot;Endothelial&quot;
#&gt; Calculating cluster Endothelial_ACKR1_HEVs
#&gt; Calculating cluster Endothelial_ESM1_tip
#&gt; Calculating cluster Endothelial_CA4_capillary
#&gt; Calculating cluster Endothelial_FBLN5_arterial
#&gt; Calculating cluster Endothelial_PROX1_lymphatic
#&gt; [1] &quot;T-CD4&quot;
#&gt; Calculating cluster T-CD4-naive_CCR7
#&gt; Calculating cluster T-CD4_PDCD1
#&gt; Calculating cluster T-CD4-TRM_JUN
#&gt; Calculating cluster T-CD4_IL17A
#&gt; [1] &quot;calculate mean and variance of reference profile&quot;
#&gt; Creating Relative Abudance Matrix...
#&gt; Creating Variance Matrix...
#&gt; Creating Relative Abudance Matrix...
#&gt; Creating Variance Matrix...
summary(d)
#&gt;               Length Class      Mode   
#&gt; majorsc.basis      2 -none-     list   
#&gt; minorsc.basis      2 -none-     list   
#&gt; bulk          336868 -none-     numeric
#&gt; markers            7 data.frame list
</code></pre>

<p>Besides the datasets generated by prepareInput(), ProM also needs the prior distribution of the cell frequency (used in simulation-based marker identification). Such information is provided to the ProM algorithm in a sample by cell frequency matrix, and the number of samples should be more than one. The samples do not need to be the same as in the input bulk or scRNAseq dataset. The most straightforward way to obtain such information is to compute it from the scRNAseq dataset using function getcellfreqM() as below, but it can also be provided based on independent assays, such as Cytometry readouts.   </p>

<pre><code class="r">priorcellfreqM&lt;-getcellfreqM(sc,clusterlabel=&quot;minorlabel&quot;)
dim(priorcellfreqM)
#&gt; [1] 11 26
priorcellfreqM[,&quot;SI_651&quot;]
#&gt;                       Basal      Endothelial_ACKR1_HEVs 
#&gt;                 0.022346369                 0.069832402 
#&gt;   Endothelial_CA4_capillary        Endothelial_ESM1_tip 
#&gt;                 0.153631285                 0.058659218 
#&gt;  Endothelial_FBLN5_arterial Endothelial_PROX1_lymphatic 
#&gt;                 0.265363128                 0.058659218 
#&gt;                     Luminal                 T-CD4_IL17A 
#&gt;                 0.008379888                 0.094972067 
#&gt;                 T-CD4_PDCD1            T-CD4-naive_CCR7 
#&gt;                 0.128491620                 0.041899441 
#&gt;               T-CD4-TRM_JUN 
#&gt;                 0.097765363
</code></pre>

<p>Finally, we can start to run ProM algorithm.</p>

<pre><code class="r">ProMEstimate&lt;-ProM(bulk=d$bulk,sc.basis=d$minorsc.basis,markers=d$markers,priorcellfreqM=priorcellfreqM)
#&gt; [1] &quot;estimation from profile&quot;
#&gt; [1] &quot;estimation from profile with bulkSigma&quot;
#&gt; [1] &quot;14 paired samples used to get bulksigma&quot;
#&gt; [1] &quot;estimation from marker&quot;
#&gt; [1] &quot;simulation without bulkSigma&quot;
#&gt; [1] &quot;estimation from markers: mean&quot;
#&gt; [1] &quot;simulation with bulkSigma&quot;
#&gt; [1] &quot;14 paired samples used to get bulksigma&quot;
#&gt; [1] &quot;estimation from markers: mean&quot;
#&gt; [1] &quot;combine estimation from profile and marker&quot;
#&gt; [1] &quot;normalizing cell frequency&quot;
#&gt; [1] &quot;The following celltype needs ProM with markerintercept Basal Endothelial_ESM1_tip Endothelial_PROX1_lymphatic Luminal T-CD4_PDCD1&quot;
#&gt; [1] &quot;estimation from profile&quot;
#&gt; [1] &quot;estimation from profile with bulkSigma&quot;
#&gt; [1] &quot;14 paired samples used to get bulksigma&quot;
#&gt; [1] &quot;estimation from marker&quot;
#&gt; [1] &quot;simulation without bulkSigma&quot;
#&gt; [1] &quot;estimation from markers: mean&quot;
#&gt; [1] &quot;simulation with bulkSigma&quot;
#&gt; [1] &quot;14 paired samples used to get bulksigma&quot;
#&gt; [1] &quot;estimation from markers: mean&quot;
#&gt; [1] &quot;combine estimation from profile and marker&quot;
#&gt; [1] &quot;normalizing cell frequency&quot;
dim(ProMEstimate)
#&gt; [1] 11 14
ProMEstimate[,1:5]
#&gt;                                   21753       21783      22144      22171
#&gt; Basal                       0.088211974 0.151577625 0.05883133 0.46885758
#&gt; Endothelial_ACKR1_HEVs      0.115466500 0.048067249 0.05195129 0.06356913
#&gt; Endothelial_CA4_capillary   0.044781289 0.013438052 0.04955488 0.01638050
#&gt; Endothelial_ESM1_tip        0.007746926 0.012109651 0.00000000 0.01706958
#&gt; Endothelial_FBLN5_arterial  0.124703865 0.048778945 0.08415359 0.06503558
#&gt; Endothelial_PROX1_lymphatic 0.113800718 0.026624597 0.04726712 0.07890376
#&gt; Luminal                     0.371659634 0.609943528 0.59272698 0.19363891
#&gt; T-CD4_IL17A                 0.031837267 0.033155558 0.04311314 0.02687055
#&gt; T-CD4_PDCD1                 0.038005018 0.007941777 0.02776161 0.01969295
#&gt; T-CD4-naive_CCR7            0.045330898 0.032543138 0.01868309 0.02717333
#&gt; T-CD4-TRM_JUN               0.018455913 0.015819881 0.02595697 0.02280812
#&gt;                                   22605
#&gt; Basal                       0.194754199
#&gt; Endothelial_ACKR1_HEVs      0.028532054
#&gt; Endothelial_CA4_capillary   0.026469430
#&gt; Endothelial_ESM1_tip        0.013865340
#&gt; Endothelial_FBLN5_arterial  0.077174863
#&gt; Endothelial_PROX1_lymphatic 0.125585387
#&gt; Luminal                     0.453038335
#&gt; T-CD4_IL17A                 0.010820759
#&gt; T-CD4_PDCD1                 0.044737603
#&gt; T-CD4-naive_CCR7            0.015475776
#&gt; T-CD4-TRM_JUN               0.009546254
</code></pre>

<p>We also provided a wrap-up function for a few other existing deconvolution methods using the same input data as computed above.The methods include here are &ldquo;CIBERSORT&rdquo;,&ldquo;DSA&rdquo;,&ldquo;DeconRNASeq&rdquo;,&ldquo;nnls&rdquo;,&ldquo;MuSiC&rdquo; and &ldquo;BisqueRNA&rdquo;.</p>

<pre><code class="r">otherEstimate&lt;-otherDeconvolution(bulk=d$bulk,sc.basis=d$minorsc.basis,markers=d$markers,sc=sc,method=&quot;MuSiC&quot;)
#&gt; [1] &quot;1795 genes included in the deconvolution&quot;
#&gt; Loading required package: xbioc
#&gt; Loading required package: Biobase
#&gt; Loading required package: BiocGenerics
#&gt; Loading required package: parallel
#&gt; 
#&gt; Attaching package: &#39;BiocGenerics&#39;
#&gt; The following objects are masked from &#39;package:parallel&#39;:
#&gt; 
#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,
#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB
#&gt; The following objects are masked from &#39;package:stats&#39;:
#&gt; 
#&gt;     IQR, mad, sd, var, xtabs
#&gt; The following objects are masked from &#39;package:base&#39;:
#&gt; 
#&gt;     anyDuplicated, append, as.data.frame, basename, cbind,
#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter,
#&gt;     Find, get, grep, grepl, intersect, is.unsorted, lapply, Map,
#&gt;     mapply, match, mget, order, paste, pmax, pmax.int, pmin,
#&gt;     pmin.int, Position, rank, rbind, Reduce, rownames, sapply,
#&gt;     setdiff, sort, table, tapply, union, unique, unsplit, which,
#&gt;     which.max, which.min
#&gt; Welcome to Bioconductor
#&gt; 
#&gt;     Vignettes contain introductory material; view with
#&gt;     &#39;browseVignettes()&#39;. To cite Bioconductor, see
#&gt;     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.
#&gt; Loading required package: AnnotationDbi
#&gt; Loading required package: stats4
#&gt; Loading required package: IRanges
#&gt; Loading required package: S4Vectors
#&gt; 
#&gt; Attaching package: &#39;S4Vectors&#39;
#&gt; The following object is masked from &#39;package:base&#39;:
#&gt; 
#&gt;     expand.grid
otherEstimate[,1:5]
#&gt;                                   21753       21783       22144
#&gt; Luminal                     0.422512476 0.507041303 0.496742644
#&gt; Endothelial_ACKR1_HEVs      0.080076968 0.010677329 0.011061227
#&gt; Endothelial_ESM1_tip        0.000000000 0.042462141 0.000000000
#&gt; Endothelial_CA4_capillary   0.120172711 0.004474830 0.123641120
#&gt; Endothelial_FBLN5_arterial  0.152591361 0.021988500 0.163556720
#&gt; Basal                       0.005435916 0.330640431 0.000000000
#&gt; Endothelial_PROX1_lymphatic 0.019416171 0.003933482 0.005942967
#&gt; T-CD4-naive_CCR7            0.000000000 0.000000000 0.000000000
#&gt; T-CD4_PDCD1                 0.199794396 0.078781984 0.199055323
#&gt; T-CD4-TRM_JUN               0.000000000 0.000000000 0.000000000
#&gt; T-CD4_IL17A                 0.000000000 0.000000000 0.000000000
#&gt;                                   22171      22605
#&gt; Luminal                     0.027977490 0.25796833
#&gt; Endothelial_ACKR1_HEVs      0.001659359 0.00000000
#&gt; Endothelial_ESM1_tip        0.000000000 0.08836864
#&gt; Endothelial_CA4_capillary   0.037407321 0.00000000
#&gt; Endothelial_FBLN5_arterial  0.159628446 0.04490338
#&gt; Basal                       0.608831250 0.17150258
#&gt; Endothelial_PROX1_lymphatic 0.017189185 0.08403036
#&gt; T-CD4-naive_CCR7            0.000000000 0.11073158
#&gt; T-CD4_PDCD1                 0.147306948 0.24249512
#&gt; T-CD4-TRM_JUN               0.000000000 0.00000000
#&gt; T-CD4_IL17A                 0.000000000 0.00000000
</code></pre>

</body>

</html>
