---
title: "introduction"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, message=FALSE}
library(ProM)
library(Seurat)
```

This R package contains ProM algorithm and wrap-up of a few other existing deconvolution algorithms for comparison purpose. The input of the ProM algorithm is a scRNAseq dataset (Seurat object) and the bulk RNAseq data in TPM measurement(gene by sample matrix). The R package contains toy examples of input datasets which can be checked as following.
```{r}
data(sc)
class(sc)
dim(sc)
data(bulk)
class(bulk)
dim(bulk)
```
For each cell in Seurat object, three input features are needed: the minor and major cell cluster each cell was assigned to, and the sample where each cell comes from. The minor cell clusters are the ones that ProM will deconvolute the bulk RNAseq into. The major cell cluster annotations are needed to facilitate ProM to derive cell markers for similar minor clusters. It is fine if some major cell clusters only consist of one minor cell cluster.

```{r}
table(sc$majorlabel)
table(sc$minorlabel)
table(sc$sampleID)
```

The first step is to use prepareInput() function to preprocess the above input data to generate needed data for the ProM deconvolution function. It includes getting the common genes between bulk and single cell datasets, renormalization, getting markers for each cell cluster, calculating reference profiles with gene mean and variance. This may take a few minutes.

```{r}
d<-prepareInput(bulk,sc,majorlabel = "majorlabel", minorlabel = "minorlabel", samplelabel = "sampleID")
summary(d)
```

Besides the datasets generated by prepareInput(), ProM also needs the prior distribution of the cell frequency (used in simulation-based marker identification). Such information is provided to the ProM algorithm in a sample by cell frequency matrix, and the number of samples should be more than one. The samples do not need to be the same as in the input bulk or scRNAseq dataset. The most straightforward way to obtain such information is to compute it from the scRNAseq dataset using function getcellfreqM() as below, but it can also be provided based on independent assays, such as Cytometry readouts.   

```{r}
priorcellfreqM<-getcellfreqM(sc,clusterlabel="minorlabel")
dim(priorcellfreqM)
priorcellfreqM[,"SI_651"]
```
Finally, we can start to run ProM algorithm.
```{r}
ProMEstimate<-ProM(bulk=d$bulk,sc.basis=d$minorsc.basis,markers=d$markers,priorcellfreqM=priorcellfreqM)
dim(ProMEstimate)
ProMEstimate[,1:5]
```

We also provided a wrap-up function for a few other existing deconvolution methods using the same input data as computed above.The methods include here are "CIBERSORT","DSA","DeconRNASeq","nnls","MuSiC" and "BisqueRNA".

```{r}
otherEstimate<-otherDeconvolution(bulk=d$bulk,sc.basis=d$minorsc.basis,markers=d$markers,sc=sc,method="MuSiC")
otherEstimate[,1:5]
```